File: source/go/lcs.go
package main

import "fmt"

func lcs(a, b string) (int, string) {
	m, n := len(a), len(b)
	dp := make([][]int, m+1)
	for i := 0; i <= m; i++ {
		dp[i] = make([]int, n+1)
	}

	for i := 1; i <= m; i++ {
		for j := 1; j <= n; j++ {
			if a[i-1] == b[j-1] {
				dp[i][j] = dp[i-1][j-1] + 1
			} else {
				if dp[i-1][j] >= dp[i][j-1] {
					dp[i][j] = dp[i-1][j]
				} else {
					dp[i][j] = dp[i][j-1]
				}
			}
		}
	}

	result := ""
	i, j := m, n
	for i > 0 && j > 0 {
		if a[i-1] == b[j-1] {
			result = string(a[i-1]) + result
			i--
			j--
		} else if dp[i-1][j] >= dp[i][j-1] {
			i--
		} else {
			j--
		}
	}

	return dp[m][n], result
}

func main() {
	a := "ABCBDAB"
	b := "BDCAB"
	length, subseq := lcs(a, b)
	fmt.Printf("lcs length: %d\n", length)
	fmt.Printf("lcs: %s\n", subseq)
}

File: source/rust/lcs.rs
fn lcs(a: &str, b: &str) -> (usize, String) {
    let a: Vec<char> = a.chars().collect();
    let b: Vec<char> = b.chars().collect();
    let m = a.len();
    let n = b.len();
    let mut dp = vec![vec![0usize; n + 1]; m + 1];

    for i in 1..=m {
        for j in 1..=n {
            if a[i-1] == b[j-1] {
                dp[i][j] = dp[i-1][j-1] + 1;
            } else {
                dp[i][j] = dp[i-1][j].max(dp[i][j-1]);
            }
        }
    }

    let mut result = String::new();
    let mut i = m;
    let mut j = n;
    while i > 0 && j > 0 {
        if a[i-1] == b[j-1] {
            result.push(a[i-1]);
            i -= 1; j -= 1;
        } else if dp[i-1][j] >= dp[i][j-1] {
            i -= 1;
        } else {
            j -= 1;
        }
    }
    let result: String = result.chars().rev().collect();
    (dp[m][n], result)
}

fn main() {
    let (length, subseq) = lcs("ABCBDAB", "BDCAB");
    println!("lcs length: {}", length);
    println!("lcs: {}", subseq);
}
